reactive/kubernetes_worker.py

```diff
@@ -242,6 +242,35 @@ def shutdown():
     service_stop('snap.kube-proxy.daemon')
 
 
+@when_not('config.changed.ports-to-open')
+def open_close_ports_if_needed():
+    # 1) Check which ports are open and save to already_opened
+    # opened_ports return format: "icmp" or "80/tcp"
+    already_opened = set([p.split("/")[0] for p in hookenv.opened_ports()])
+    # 2) iterate over the list from ports-to-open and check if not present to alrady_opened
+    to_open = hookenv.config().get("ports-to-open")
+    if to_open is None:
+        to_open = ""
+    # Basic clean-up: remove any dups with set()
+    # clean-up string, breaking on , and replacing empty spaces
+    to_open_set = set([p.replace(" ","") for p in to_open.split(",")])
+    for p in to_open_set:
+        if not p.isdigit():
+            hookenv.log("[WARN] ports-to-open: port {} must have only digits, ignoring...".format(p))
+        if p not in already_opened:
+            try:
+                hookenv.open_port(p)
+            except Exception as e: # open_port may fail because port has been opened by another charm, for example; just log it
+                hookenv.log("[WARN] ports-to-open failed to open {}, could that be another charm or application holding it open?".format(p))
```

Comment:
This is a tough position. If it's opened by another charm and the operator is just adding it here for clarity it isn't an error, but if they request a port to be opened and it fails for some reason I think we should notify the operator somewhere other than just the log. Is the already open error a unique one that we can single out and log only?

Reply:
I cannot see on Juju's code where open-ports is actually defined. I think this is a very complex question because it will change behavior per provider. On OpenStack, for example, open-ports means adding rules to security groups. There are only two situations I see it failing in this case: (1) we have no rights to edit sec groups, but then it should also fail on bootstrap; or (2) a connection timeout, which is an actual problem.
Whereas vsphere always fails with NotImplemented.
https://github.com/juju/juju/blob/a5ab92ec9b7f5da3678d9ac603fe52d45af24412/provider/vsphere/environ_network.go#L14


To be honest, I believe we have only the logging option here, because if we raise it as an exception, it will force the user to scout around the instance and other charms until finding which is causing problem (or try it by himself).
We will also need to triage according to provider (we may have this info via integrator, though).